# Project Code

In [1]:
# import libraries
import numpy as np
import pandas as pd
import altair as alt
from sklearn.decomposition import PCA

In [2]:
# importing data and printing first 5 rows
data_mod = pd.read_csv('2015.csv')
data_mod.head()

Country          Region  Happiness Rank  Happiness Score  \
0  Switzerland  Western Europe               1            7.587   
1      Iceland  Western Europe               2            7.561   
2      Denmark  Western Europe               3            7.527   
3       Norway  Western Europe               4            7.522   
4       Canada   North America               5            7.427   

   Standard Error  Economy (GDP per Capita)   Family  \
0         0.03411                   1.39651  1.34951   
1         0.04884                   1.30232  1.40223   
2         0.03328                   1.32548  1.36058   
3         0.03880                   1.45900  1.33095   
4         0.03553                   1.32629  1.32261   

   Health (Life Expectancy)  Freedom  Trust (Government Corruption)  \
0                   0.94143  0.66557                        0.41978   
1                   0.94784  0.62877                        0.14145   
2                   0.87464  0.64938                        0.48357   
3                   0.88521  0.66973                        0.36503   
4                   0.90563  0.63297                        0.32957   

   Generosity  Dystopia Residual  
0     0.29678            2.51738  
1     0.43630            2.70201  
2     0.34139            2.49204  
3     0.34699            2.46531  
4     0.45811            2.45176

In [3]:
# renaming columns
rename_var = {'Country': 'country',
              'Region': 'region',
              'Happiness Rank': 'hap_rank',
              'Happiness Score': 'hap_score',
              'Standard Error': 'std_err',
              'Economy (GDP per Capita)': 'econ',
              'Family': 'fam',
              'Health (Life Expectancy)': 'health',
              'Freedom': 'freedom',
              'Trust (Government Corruption)': 'gov_trust',
              'Generosity': 'generosity',
              'Dystopia Residual': 'dys_res'
             }

data = data_mod.rename(columns=rename_var)

In [4]:
# printing first 5 rows of tidied dataset
data.head(5)

country          region  hap_rank  hap_score  std_err     econ  \
0  Switzerland  Western Europe         1      7.587  0.03411  1.39651   
1      Iceland  Western Europe         2      7.561  0.04884  1.30232   
2      Denmark  Western Europe         3      7.527  0.03328  1.32548   
3       Norway  Western Europe         4      7.522  0.03880  1.45900   
4       Canada   North America         5      7.427  0.03553  1.32629   

       fam   health  freedom  gov_trust  generosity  dys_res  
0  1.34951  0.94143  0.66557    0.41978     0.29678  2.51738  
1  1.40223  0.94784  0.62877    0.14145     0.43630  2.70201  
2  1.36058  0.87464  0.64938    0.48357     0.34139  2.49204  
3  1.33095  0.88521  0.66973    0.36503     0.34699  2.46531  
4  1.32261  0.90563  0.63297    0.32957     0.45811  2.45176

In [5]:
# exporting tidied dataset
data.to_csv('tidy-data.csv', index=False)

In [6]:
# check dimension of data set
data.shape

(158, 12)

##### The dimensions of our 2015 World Happiness Report datsset consists of 158 rows and 12 columns.

In [7]:
# check for null values
data.isnull().sum()

country       0
region        0
hap_rank      0
hap_score     0
std_err       0
econ          0
fam           0
health        0
freedom       0
gov_trust     0
generosity    0
dys_res       0
dtype: int64

In [8]:
# check for 0.0 values
(data == 0).sum()

country       0
region        0
hap_rank      0
hap_score     0
std_err       0
econ          1
fam           1
health        1
freedom       1
gov_trust     1
generosity    1
dys_res       0
dtype: int64

In [9]:
# check whether a row is missing many values
data.loc[(data == 0.0).any(axis=1)]

country                           region  hap_rank  \
73                  Indonesia                Southeastern Asia        74   
101                    Greece                   Western Europe       102   
111                      Iraq  Middle East and Northern Africa       112   
119          Congo (Kinshasa)               Sub-Saharan Africa       120   
122              Sierra Leone               Sub-Saharan Africa       123   
147  Central African Republic               Sub-Saharan Africa       148   

     hap_score  std_err     econ      fam   health  freedom  gov_trust  \
73       5.399  0.02596  0.82827  1.08708  0.63793  0.46611    0.00000   
101      4.857  0.05062  1.15406  0.92933  0.88213  0.07699    0.01397   
111      4.677  0.05232  0.98549  0.81889  0.60237  0.00000    0.13788   
119      4.517  0.03680  0.00000  1.00120  0.09806  0.22605    0.07625   
122      4.507  0.07068  0.33024  0.95571  0.00000  0.40840    0.08786   
147      3.678  0.06112  0.07850  0.00000  0.06699  0.48879    0.08289   

     generosity  dys_res  
73      0.51535  1.86399  
101     0.00000  1.80101  
111     0.17922  1.95335  
119     0.24834  2.86712  
122     0.21488  2.51009  
147     0.23835  2.72230

##### There are no missing values in any columns.

In [10]:
# Variable summary for quantitative columns in dataset 
data_columns = ['hap_rank',
                'std_err',
                'econ',
                'fam',
                'health',
                'gov_trust',
                'generosity',
                'dys_res'
               ]

column_summary = {'Quantitative Columns': data_columns,
                  'Min': data[data_columns].min().values,
                  'Max': data[data_columns].max().values,
                  'Mean': data[data_columns].mean().values
                 }

data_summary = pd.DataFrame(column_summary)
data_summary

Quantitative Columns      Min        Max       Mean
0             hap_rank  1.00000  158.00000  79.493671
1              std_err  0.01848    0.13693   0.047885
2                 econ  0.00000    1.69042   0.846137
3                  fam  0.00000    1.40223   0.991046
4               health  0.00000    1.02525   0.630259
5            gov_trust  0.00000    0.55191   0.143422
6           generosity  0.00000    0.79588   0.237296
7              dys_res  0.32858    3.60214   2.098977

In [11]:
data.region.value_counts()

Sub-Saharan Africa                 40
Central and Eastern Europe         29
Latin America and Caribbean        22
Western Europe                     21
Middle East and Northern Africa    20
Southeastern Asia                   9
Southern Asia                       7
Eastern Asia                        6
Australia and New Zealand           2
North America                       2
Name: region, dtype: int64

### Exploratory Analysis

In [12]:
# correlation
# store qualitative variables and unrelated quantitative variables separately 
x_mx = data.drop(columns = ['country', 'region', 'hap_rank', 'std_err'])

In [13]:
# correlation matrix
corr_mx = x_mx.corr()

In [14]:
# correlation between all variables and Happiness Score
corr_mx.loc[:, 'hap_score'].sort_values() 

generosity    0.180319
gov_trust     0.395199
dys_res       0.530474
freedom       0.568211
health        0.724200
fam           0.740605
econ          0.780966
hap_score     1.000000
Name: hap_score, dtype: float64

In [15]:
# melt corr_mx
corr_mx_long = corr_mx.reset_index().rename(
    columns = {'index': 'row'}
).melt(
    id_vars = 'row',
    var_name = 'col',
    value_name = 'Correlation'
)

# construct plot
corr_heatmap = alt.Chart(corr_mx_long).mark_rect().encode(
    x = alt.X('col', title = '', sort = {'field': 'Correlation', 'order': 'ascending'}), 
    y = alt.Y('row', title = '', sort = {'field': 'Correlation', 'order': 'ascending'}),
    color = alt.Color('Correlation', 
                      scale = alt.Scale(scheme = 'blueorange', # diverging gradient
                                        domain = (-1, 1), # ensure white = 0
                                        type = 'sqrt'), # adjust gradient scale
                     legend = alt.Legend(tickCount = 5)) # add ticks to colorbar at 0.5 for reference
).properties(width = 300, height = 300)

corr_heatmap

alt.Chart(...)

In [16]:
# remove happiness score for PCA
x_mx_pca = x_mx.drop(columns = ['hap_score'])

In [17]:
# PCA
# center and scale ('normalize')
x_ctr = (x_mx_pca - x_mx_pca.mean())/x_mx_pca.std()

In [18]:
# compute principal components
pca = PCA(n_components = x_ctr.shape[1]) 
pca.fit(x_ctr)

PCA(n_components=7)

In [19]:
# store proportion of variance explained as a dataframe
pca_var_explained = pd.DataFrame({'Proportion of variance explained': pca.explained_variance_ratio_})

# add component number as a new column
pca_var_explained['Component'] = np.arange(1, 8)

# add cumulative variance explained as a new column
pca_var_explained['Cumulative variance explained'] = np.cumsum(pca_var_explained.loc[:, 'Proportion of variance explained'].values, axis = 0)

# print
pca_var_explained.head()

Proportion of variance explained  Component  Cumulative variance explained
0                          0.411648          1                       0.411648
1                          0.194166          2                       0.605814
2                          0.143014          3                       0.748828
3                          0.100409          4                       0.849237
4                          0.074336          5                       0.923572

In [20]:
# create dual-axis plot to determine number of PCs
# encode component axis only as base layer
base = alt.Chart(pca_var_explained).encode(
    x = 'Component')

# make a base layer for the proportion of variance explained
prop_var_base = base.encode(
    y = alt.Y('Proportion of variance explained',
              axis = alt.Axis(titleColor = '#57A44C'))
)

# make a base layer for the cumulative variance explained
cum_var_base = base.encode(
    y = alt.Y('Cumulative variance explained', axis = alt.Axis(titleColor = '#5276A7'))
)

# add points and lines to each base layer
prop_var = prop_var_base.mark_line(stroke = '#57A44C') + prop_var_base.mark_point(color = '#57A44C')
cum_var = cum_var_base.mark_line() + cum_var_base.mark_point()

# layer the layers
var_explained_plot = alt.layer(prop_var, cum_var).resolve_scale(y = 'independent')

# display
var_explained_plot

alt.LayerChart(...)

In [21]:
# find number of PCs that captures more than 15% of variance 
main_pca = sum(pca_var_explained['Proportion of variance explained'] > 0.15)

#print
main_pca

2

In [22]:
# store the loadings as a data frame with appropriate names
loading_df = pd.DataFrame(pca.components_).transpose().rename(
    columns = {0: 'PC1', 1: 'PC2'} # add entries for each selected component
).loc[:, ['PC1', 'PC2']] # slice just components of interest

# add a column with the variable names
loading_df['Variable'] = x_mx_pca.columns.values
# print
loading_df

PC1       PC2    Variable
0 -0.499316 -0.311352        econ
1 -0.452652 -0.231072         fam
2 -0.478673 -0.242881      health
3 -0.418860  0.339110     freedom
4 -0.328969  0.418443   gov_trust
5 -0.174669  0.639575  generosity
6 -0.049877 -0.302536     dys_res

In [23]:
# melt from wide to long
loading_plot_df = loading_df.melt(
    id_vars = 'Variable',
    var_name = 'Principal Component',
    value_name = 'Loading'
)

# add a column of zeros to encode for x = 0 line to plot
loading_plot_df['zero'] = np.repeat(0, len(loading_plot_df))

# create base layer
base = alt.Chart(loading_plot_df)

# create lines + points for loadings
loadings = base.mark_line(point = True).encode(
    y = alt.X('Variable', title = ''),
    x = 'Loading',
    color = 'Principal Component'
)

# create line at zero
rule = base.mark_rule().encode(x = alt.X('zero', title = 'Loading'), size = alt.value(0.05))

# layer
loading_plot = (loadings + rule).properties(width = 120)

# show
loading_plot.facet('Principal Component')

alt.FacetChart(...)

In [24]:
# Scatterplot for econ
fig_econ = alt.Chart(data).mark_circle(opacity = 0.5).encode(
    x = alt.X('econ', title = 'Economy (GDP)'),
    y = alt.Y('hap_score', title = 'Happiness Score', scale = alt.Scale(type = 'log')),
    color = alt.Color('region', title = 'Region')).properties(width = 250, height = 250)

fig_econ

alt.Chart(...)

In [25]:
# Scatterplot for fam
fig_fam = alt.Chart(data).mark_circle(opacity = 0.5).encode(
    x = alt.X('fam', title = 'Family'),
    y = alt.Y('hap_score', title = 'Happiness Score', scale = alt.Scale(type = 'log')),
    color = alt.Color('region', title = 'Region')).properties(width = 250, height = 250)

fig_fam

alt.Chart(...)

In [26]:
# Scatterplot for health
fig_health = alt.Chart(data).mark_circle(opacity = 0.5).encode(
    x = alt.X('health', title = 'Health'),
    y = alt.Y('hap_score', title = 'Happiness Score', scale = alt.Scale(type = 'log')),
    color = alt.Color('region', title = 'Region')).properties(width = 400, height = 400)

fig_health

alt.Chart(...)

In [27]:
# Scatterplot for freedom
fig_freedom = alt.Chart(data).mark_circle(opacity = 0.5).encode(
    x = alt.X('freedom', title = 'Freedom'),
    y = alt.Y('hap_score', title = 'Happiness Score', scale = alt.Scale(type = 'log')),
    color = alt.Color('region', title = 'Region')).properties(width = 400, height = 400)

fig_freedom

alt.Chart(...)

In [28]:
# Scatterplot for gov_trust
fig_gov = alt.Chart(data).mark_circle(opacity = 0.5).encode(
    x = alt.X('gov_trust', title = 'Government Trust'),
    y = alt.Y('hap_score', title = 'Happiness Score', scale = alt.Scale(type = 'log')),
    color = alt.Color('region', title = 'Region')).properties(width = 400, height = 400)

fig_gov

alt.Chart(...)

In [29]:
# Scatterplot for generosity
fig_generosity = alt.Chart(data).mark_circle(opacity = 0.5).encode(
    x = alt.X('generosity', title = 'Generosity'),
    y = alt.Y('hap_score', title = 'Happiness Score', scale = alt.Scale(type = 'log')),
    color = alt.Color('region', title = 'Region')).properties(width = 400, height = 400)

fig_generosity

alt.Chart(...)

In [30]:
fig_econ | fig_fam

alt.HConcatChart(...)

### Regression Analysis

In [31]:
from sklearn.linear_model import LinearRegression
from sklearn.metrics import r2_score
from sklearn.preprocessing import add_dummy_feature

In [32]:
x_mx

hap_score     econ      fam   health  freedom  gov_trust  generosity  \
0        7.587  1.39651  1.34951  0.94143  0.66557    0.41978     0.29678   
1        7.561  1.30232  1.40223  0.94784  0.62877    0.14145     0.43630   
2        7.527  1.32548  1.36058  0.87464  0.64938    0.48357     0.34139   
3        7.522  1.45900  1.33095  0.88521  0.66973    0.36503     0.34699   
4        7.427  1.32629  1.32261  0.90563  0.63297    0.32957     0.45811   
..         ...      ...      ...      ...      ...        ...         ...   
153      3.465  0.22208  0.77370  0.42864  0.59201    0.55191     0.22628   
154      3.340  0.28665  0.35386  0.31910  0.48450    0.08010     0.18260   
155      3.006  0.66320  0.47489  0.72193  0.15684    0.18906     0.47179   
156      2.905  0.01530  0.41587  0.22396  0.11850    0.10062     0.19727   
157      2.839  0.20868  0.13995  0.28443  0.36453    0.10731     0.16681   

     dys_res  
0    2.51738  
1    2.70201  
2    2.49204  
3    2.46531  
4    2.45176  
..       ...  
153  0.67042  
154  1.63328  
155  0.32858  
156  1.83302  
157  1.56726  

[158 rows x 8 columns]

In [33]:
# Store happiness score (response variable) as array
y = x_mx.hap_score

# Store explanatory variable
exp_names = ['econ', 'fam', 'health', 'freedom', 'gov_trust', 'generosity', 'dys_res']
x_slr_df = x_mx.loc[:, exp_names]

# add intercept column
x_slr = add_dummy_feature(x_slr_df, value = 1)

# print first 5 rows
x_slr[0:5, :]

array([[1.     , 1.39651, 1.34951, 0.94143, 0.66557, 0.41978, 0.29678,
        2.51738],
       [1.     , 1.30232, 1.40223, 0.94784, 0.62877, 0.14145, 0.4363 ,
        2.70201],
       [1.     , 1.32548, 1.36058, 0.87464, 0.64938, 0.48357, 0.34139,
        2.49204],
       [1.     , 1.459  , 1.33095, 0.88521, 0.66973, 0.36503, 0.34699,
        2.46531],
       [1.     , 1.32629, 1.32261, 0.90563, 0.63297, 0.32957, 0.45811,
        2.45176]])

#### Model Fitting

In [34]:
# configure module
slr = LinearRegression(fit_intercept = False)

# fit model
slr.fit(x_slr, y)

LinearRegression(fit_intercept=False)

In [35]:
#store the coefficient estimates
coef = slr.coef_

# print coefficient estimates
coef

array([6.40485915e-05, 1.00010140e+00, 9.99970348e-01, 9.99882614e-01,
       9.99695308e-01, 9.99919136e-01, 1.00006126e+00, 1.00003038e+00])

In [36]:
# store dimensions
n, p = x_slr.shape

# compute x'x
xtx = x_slr.transpose().dot(x_slr)

# compute x'x inverse
xtx_inv = np.linalg.inv(xtx)

# compute residuals
fitted_slr = slr.predict(x_slr)
resid = y - fitted_slr

# compute error variance estimate
sigmasqhat = resid.var()*(n-1)/(n-p)

# compute variance-covariance matrix
v_hat = xtx_inv * sigmasqhat

# compute standard errors
coef_se = np.sqrt(v_hat.diagonal())
coef_se = np.append(coef_se, float('nan'))

# coefficient labels
coef_labels = np.append('intercept', exp_names)
coef_labels = np.append(coef_labels, 'error_variance')
# estimates
coef_estimates = np.append(slr.coef_, sigmasqhat)
# summary table
coef_table = pd.DataFrame(
    data = {'coef_estimates': coef_estimates, 'coef_se': coef_se},
    index = coef_labels
)

# print
print(coef_table)

                coef_estimates   coef_se
intercept         6.404859e-05  0.000125
econ              1.000101e+00  0.000113
fam               9.999703e-01  0.000115
health            9.998826e-01  0.000162
freedom           9.996953e-01  0.000198
gov_trust         9.999191e-01  0.000224
generosity        1.000061e+00  0.000202
dys_res           1.000030e+00  0.000042
error_variance    7.957829e-08       NaN


### ANOVA

In [37]:
import statsmodels.api as sm
from statsmodels.formula.api import ols

# Ordinary Least Squares (OLS) model
model = ols('hap_score ~ econ + fam + health + freedom + gov_trust + generosity + dys_res', data=data).fit()
anova_table = sm.stats.anova_lm(model)
anova_table

df      sum_sq       mean_sq             F  PR(>F)
econ          1.0  125.539976  1.255400e+02  1.577566e+09     0.0
fam           1.0   19.752378  1.975238e+01  2.482132e+08     0.0
health        1.0    4.444964  4.444964e+00  5.585650e+07     0.0
freedom       1.0    8.751990  8.751990e+00  1.099796e+08     0.0
gov_trust     1.0    1.192792  1.192792e+00  1.498891e+07     0.0
generosity    1.0    0.300449  3.004490e-01  3.775515e+06     0.0
dys_res       1.0   45.852008  4.585201e+01  5.761874e+08     0.0
Residual    150.0    0.000012  7.957829e-08           NaN     NaN

## Plots

In [38]:
data_label = data
# labels for the factor
hap_labels = ['1_low', '2_medium', '3_high']

# cut hap_fac based on quantiles and store as new variable
data_label['hap_fac'] = pd.qcut(data_label.hap_score, 
                              q = 3, 
                              labels = hap_labels)

# preview
data_label.sort_values('country').head(5)

country                           region  hap_rank  hap_score  \
152  Afghanistan                    Southern Asia       153      3.575   
94       Albania       Central and Eastern Europe        95      4.959   
67       Algeria  Middle East and Northern Africa        68      5.605   
136       Angola               Sub-Saharan Africa       137      4.033   
29     Argentina      Latin America and Caribbean        30      6.574   

     std_err     econ      fam   health  freedom  gov_trust  generosity  \
152  0.03084  0.31982  0.30285  0.30335  0.23414    0.09719     0.36510   
94   0.05013  0.87867  0.80434  0.81325  0.35733    0.06413     0.14272   
67   0.05099  0.93929  1.07772  0.61766  0.28579    0.17383     0.07822   
136  0.04758  0.75778  0.86040  0.16683  0.10384    0.07122     0.12344   
29   0.04612  1.05351  1.24823  0.78723  0.44974    0.08484     0.11451   

     dys_res   hap_fac  
152  1.95210     1_low  
94   1.89894  2_medium  
67   2.43209  2_medium  
136  1.94939     1_low  
29   2.83600    3_high

In [49]:
# plotting points according to country's hap_fac
# you can change the x axis to any explanatory variable of your choosing @PRISS
scatter = alt.Chart(data_label).mark_circle().encode(
    x = alt.X('generosity', title = 'Generosity'),
    y = alt.Y('hap_score', title = 'Happiness Score', scale = alt.Scale(zero = False)),
    color = alt.Color('hap_fac:N', title = 'Happiness Factor'))

scatter

alt.Chart(...)

In [50]:
# Creating facet for each of the different factors
scatter_fac = scatter.properties(width = 250, height = 200).facet(column = 'hap_fac')

scatter_fac

alt.FacetChart(...)

In [41]:
data

country                           region  hap_rank  hap_score  \
0    Switzerland                   Western Europe         1      7.587   
1        Iceland                   Western Europe         2      7.561   
2        Denmark                   Western Europe         3      7.527   
3         Norway                   Western Europe         4      7.522   
4         Canada                    North America         5      7.427   
..           ...                              ...       ...        ...   
153       Rwanda               Sub-Saharan Africa       154      3.465   
154        Benin               Sub-Saharan Africa       155      3.340   
155        Syria  Middle East and Northern Africa       156      3.006   
156      Burundi               Sub-Saharan Africa       157      2.905   
157         Togo               Sub-Saharan Africa       158      2.839   

     std_err     econ      fam   health  freedom  gov_trust  generosity  \
0    0.03411  1.39651  1.34951  0.94143  0.66557    0.41978     0.29678   
1    0.04884  1.30232  1.40223  0.94784  0.62877    0.14145     0.43630   
2    0.03328  1.32548  1.36058  0.87464  0.64938    0.48357     0.34139   
3    0.03880  1.45900  1.33095  0.88521  0.66973    0.36503     0.34699   
4    0.03553  1.32629  1.32261  0.90563  0.63297    0.32957     0.45811   
..       ...      ...      ...      ...      ...        ...         ...   
153  0.03464  0.22208  0.77370  0.42864  0.59201    0.55191     0.22628   
154  0.03656  0.28665  0.35386  0.31910  0.48450    0.08010     0.18260   
155  0.05015  0.66320  0.47489  0.72193  0.15684    0.18906     0.47179   
156  0.08658  0.01530  0.41587  0.22396  0.11850    0.10062     0.19727   
157  0.06727  0.20868  0.13995  0.28443  0.36453    0.10731     0.16681   

     dys_res hap_fac  
0    2.51738  3_high  
1    2.70201  3_high  
2    2.49204  3_high  
3    2.46531  3_high  
4    2.45176  3_high  
..       ...     ...  
153  0.67042   1_low  
154  1.63328   1_low  
155  0.32858   1_low  
156  1.83302   1_low  
157  1.56726   1_low  

[158 rows x 13 columns]

In [44]:
data.set_index('country', inplace = True)

In [120]:
low_gen = data[(data['hap_fac'] == '1_low') & (data.generosity > 0.7)].hap_score.idxmin()
low_gen

'Myanmar'

In [111]:
med_gen = data[(data['hap_fac'] == '2_medium') & (data.generosity < 0.1)].hap_score.idxmin()
med_gen

'Greece'

In [118]:
high_gen = data[(data['hap_fac'] == '3_high') & (data.generosity > 0.55)].hap_score.idxmax()
high_gen

'Thailand'

'Myanmar'

In [121]:
point_labels = data.loc[['Myanmar', 'Greece', 'Thailand', 'Mexico']].reset_index()

# plot text labels
labels = alt.Chart(point_labels).mark_text(align = 'left', dx = 6, dy = 3).encode(
    x = 'generosity',
    y = 'hap_score',
    text = 'country'
)

# layer over scatter
scatter + labels

alt.LayerChart(...)